feat(api): add v4 JSON executor

shlokamin opened the pull request:
## overview

This PR closes #4868 by adding the v4 JSON executor to the api

## changelog

  - Add support for V4 JSON schema in v4 executor
  - Move temperature module await temperature command down to module level
  - Update versioning notes
  - Updated API docs to mention only two pipettes are supported
  - Add start_set_temp and await_temp to run log

## review requests

- Code review
- Create a protocol with magnetic + temperature modules in PD with their respective steps. Upload to an OT-2 using the run app and validate that the protocol runs smoothly. 
- Test the new temperature module `start_set_temperature` and `await_temperature` commands. 
- Test both gen 1 and gen2 modules
- Make sure commands show up in run log. Note, await temp will show up as a "pause" in the run log
- ~Regression test that protocols that do not use modules still behave as expected~ Nevermind if you export a protocol without modules it will export as a v3 protocol, and therefore use the v3 executor

sfoster1 committed:
wip modules gen2

sfoster1 committed:
fixup: address comments, fix tests

sfoster1 committed:
fixup: actually fix tests

sfoster1 committed:
fixup: fix lint

sfoster1 committed:
fixup: changes to make FE lives easier

IanLondon committed:
wip: try using mocks instead of MockPipette. (based on api_refactor-v3-executor-test branch)

IanLondon committed:
flow_rate spies working with ugly verbose mocking

IanLondon committed:
unit tests, no integration-with-mock

IanLondon committed:
fix linting

IanLondon committed:
refactor to avoid passing protocol_data around for no reason

IanLondon committed:
refactor v3 executor helper fns to use minimum set of args

IanLondon committed:
execute_v4 stubs

IanLondon committed:
take executor helpers and tests as far as we can waiting on other work

IanLondon committed:
implement load_modules_from_json, start hooking v4 executor to execute.py

IanLondon committed:
Merge remote-tracking branch 'origin/edge' into api_refactor-v3-executor-test

shlokamin committed:
feat(api): add v4 JSON executor

#4868

shlokamin committed:
feat(api): add v4 JSON executor

#4868

shlokamin committed:
feat: include all vscode configs in gitignore (#5222)

shlokamin committed:
fix error string interpolation

Commit:
refactor(robot-server,api): robot server port networking fast api (#5184)

* create system/wifi module and move some functions from server/endpoints/networking there.

* more refactoring

* code clean up and restructuring.

* move eap option validation from network endpoints to wifi.
rewrote test_key_lifecycle to test the key api. not the http api.

* reworked test_networking_endpoints to only test the http interaction with the key functions. Not the underlying file system. That belongs in the test_wifi module.

* test_network_status added to nmcli. this didn't belong in the server endpoint tests as it is really testing our ability to communicate with nmcli.

* remove deprecated comment

* start robot_server networking endpoints

* progress

* keys is a reserved word. use an alias.

* list keys tests pass

* networking tests pass minus the disconnect

* disconnect tests

* tests for wificonfiguration model validation

* add todo note regarding wifi disconnect

* fix lint errors

* cannot import log_control.MAX_RECORDS due to syslog import failure.

* clean up the instances of multiple response types.

* rename V1ErrorResponse to V1BasicResponse. It isn't always used for errors. Name was misleading.

* fix bad test

* fix lint error

* fix bugs that were revealed when running on PI.
Configure didn't work because we weren't testing how the nmcli.configure call was made.
get wifi/keys failed due to validation errors on pydantic model due to attribute named keys. also, someone (me) didn't port the aiohttp unit tests properly.

closes #5143

shlokamin committed:
Merge branch 'api_refactor-v3-executor-test' into api_add-v4-json-executor

sfoster1 commented on api/src/opentrons/protocol_api/execute_v4.py:
```diff
@@ -0,0 +1,154 @@
+    # while holding, because the temperature delta will be exceeded.
+    # when this flicker happens, this function would sleep for a moment longer
+
+    if status == 'heating':
```
We should push this logic down all the way to the `hardware_control.modules.TemperatureModule`, so cancelling works here.

The way cancel is currently implemented is in the [execution manager in hardware control](https://github.com/Opentrons/opentrons/blob/edge/api/src/opentrons/hardware_control/execution_manager.py) with an [asyncio condition](https://docs.python.org/3/library/asyncio-sync.html#asyncio.Condition) wrapping a state that can be running, paused, or cancelling. When you call a hardware control function that checks the condition (or one of the whitelisted ones for cancellation), you can be cancelled.

As this code stands, the cancellation won't actually happen here - you could be awaiting for a very long time, especially if there's an error in the tempdeck communications.

Making a `tempdeck.await_temperature(target: float)` function will fix this. We can register it as cancellable with the execution manager, which means it will get cancelled when a cancel happens; or we can implement some more complex logic that pairs monitoring the condition with awaiting the temperature.

The other open question is, what's supposed to happen to the protocol state when this is running? How does it interact with pause? This is as much a UX question as a technical question (@howisthisnamenottakenyet ).

shlokamin replied:
Eek, forgot about cancel. I'm fine with moving this down to the module level, that means we're exposing another "hidden" thing to the module context though. But I guess there isn't really another option.

sfoster1 replied:
it's really of equal status to the other one though - they come as a pair.

Honestly it's reasonable at this point to unhide it - we wanted to keep it hidden because you could set the temperature, but waiting for it wasn't exposed. If properly waiting for it is exposed, we might as well have it be part of the python api. What do you think @pantslakz and @howisthisnamenottakenyet

sfoster1 commented on api/src/opentrons/protocol_api/execute_v4.py:
```diff
@@ -0,0 +1,154 @@
+            asyncio.sleep(0.2)
+
+
+dispatcher_map: Dict[Any, Any] = {
```
This map and the code in `dispatch_json` looks like it will work, but I'd prefer to avoid magic strings in favor of enums - it's a bit easier to type systematically, since we don't have literal types (yet) in mypy. In addition, python enums can have arbitrary objects as values, so you could have something like

```
class JsonCommandHandler(NamedTuple):  # could also be a typed generic passed down to the handler element (see above re: C++ brain)
   name: str
   handler: Callable[] # appropriately type

class JsonCommand:
   "delay": JsonCommandHandler(...)
```

sfoster1 commented on api/src/opentrons/protocol_api/execute_v4.py:
```diff
@@ -0,0 +1,154 @@
+
+
+def dispatch_json(context: ProtocolContext,
+                  protocol_data: Dict[Any, Any],
```
Similar to the comment above about typing, it would be nice for a lot of these arguments to be more tightly-defined objects rather than string-keyed dicts; it would also be nice for the protocol parse stage to yield data structures that actually have object references rather than string ids referring to keys in other string-keyed dicts.

sfoster1 commented on api/src/opentrons/protocol_api/module_geometry.py:
```diff
@@ -525,12 +534,15 @@ def resolve_module_model(module_name: str) -> ModuleModel:
-                         'Valid names (ignoring case): '
-                         '"' + '", "'.join(alias_map.keys()) + '"')
+                         'Valid names (ignoring case): ''"' + '", "'
+                         .join(model_map.keys()) + '", "' + '", "'
```
This is intended to be a literate error for python protocols; it may be better to avoid mentioning the `model_map` at all, or to explicitly mention it as a lesser alternative, something like:

"Valid names (ignoring case): {alias_map.keys()}

You can also refer to modules by their exact model: {model_map.keys()}"

sfoster1 commented on api/src/opentrons/protocol_api/module_contexts.py:
```diff
@@ -232,6 +232,16 @@ def target(self):
         return self._module.target
 
+    @property  # type: ignore
+    @requires_version(2, 0)
```
This should require version 2.3, since it will be new in 2.3. Also, changes or additions to api-exposed methods should be accompanied with a change note in the [version docs](https://github.com/Opentrons/opentrons/blob/edge/api/docs/v2/versioning.rst)

sfoster1 reviewed (changes requested):
Preliminary change requests to make this a bit more idiomatic python and support cancel

Kadee80 committed:
feat(protocol-designer): Add module prefix to labware (#5205)

closes #5119

IanLondon committed:
feat(protocol-designer): add modal to add a pause after set temp (#5182)

Closes #5117

* refactor add/select step selectors for mocking, and add tests

shlokamin committed:
move temperature module await temperature command down to module level

shlokamin committed:
add versioning notes

shlokamin commented on api/src/opentrons/hardware_control/modules/tempdeck.py:
```diff
@@ -189,6 +192,29 @@ def __init__(self,
+        the specified temperature is reached
+        """
+        await self.wait_for_is_running()
+
```
made `_await_temperature` a closure function to make it private

sfoster1 replied:
yeah that's fine with me. we're in general not concerned about enforcing visibility - we use the standard python practice of making private members start with underscores and trust people not to abuse it - but it's not worth changing this implementation

shlokamin committed:
format test json protocol

sfoster1 commented on api/src/opentrons/hardware_control/modules/tempdeck.py:
```diff
@@ -189,6 +192,29 @@ def __init__(self,
+
+        t = self._loop.create_task(_await_temperature(awaiting_temperature))
+        await self.make_cancellable(t)
+        return await t
```
what does this return? if it's nothing, we don't need to `return`; if it's something, we should annotate the return type in both the inner closure and the wrapper

sfoster1 commented on api/src/opentrons/hardware_control/modules/tempdeck.py:
```diff
@@ -189,6 +192,29 @@ def __init__(self,
+            status = self.status
+
+            if status == 'heating':
+                while (self.temperature < awaiting_temperature):
```
unnecessary parens

sfoster1 commented on api/src/opentrons/hardware_control/modules/tempdeck.py:
```diff
@@ -189,6 +192,29 @@ def __init__(self,
+                    asyncio.sleep(0.2)
+
+            elif status == 'cooling':
+                while (self.temperature > awaiting_temperature):
```
unnecessary parens

sfoster1 committed:
chore: add risk assessment field to pr template (#5233)

In a risk assessment, you go over your change set and identify other
areas of the system it might touch. Particularly important in a broad
and connected system, this process both gets the person submitting the
PR to consider the impact of their changes and points testers and
reviewers in the right direction.

Commit:
refactor(robot-server, api): implement module endpoints in fastapi (#5229)

* add hardware fixture

* use mock hardware for health test

* port module unit tests from aiohttp to fastapi

* move module entpoints and modules out of control and into their own python files.

* minor fixes in api

* create find_matching_module function and use it.

* use V1BasicResponse

closes #5141

shlokamin committed:
remove unecessary return statement and parens

Commit:
refactor(robot-server, api): add /pipettes endpoint and tests (#5237)

* add /pipettes endpoint and tests
use V1BasicResponse

* rename Pipette model to the more accurate PipettesByMount

closes #5145

Commit:
feat(protocol-designer): remove warnings and display correct version for gen 2 modules (#5224)

* feat(protocol-designer): remove warnings and display correct version of module image on file page

closes #5152

* remove crash warning from edit modules and add tests

* remove comment

* make crashable warning text only show for the crashable module

* fix flow error

IanLondon committed:
feat(protocol-designer): allow user to bypass typeform modal (#5169)

merge hotfix changes from PD release 3.0.8

Closes #5126

mcous committed:
build(styled-components): add styled-components to app deps (#5236)

sfoster1 committed:
fix(api): properly parse escaped nmcli responses (#5246)

nmcli uses ':' as a field separator. If it encounters a : in a field
value, it escapes it with a backslash. We weren't parsing this
correctly (or indeed at all) and therefore couldn't connect to wifi
networks whose names include a colon. Parsing escaped strings is hard,
so this change uses the csv module to parse them properly

mcous committed:
refactor(app): add client-side state for Wi-Fi disconnect (#5234)

iansolano committed:
feat(robot-server): add formatted json responses and error handling (#5208)

* feat(robot-server): add formatted json responses and error handling

SyntaxColoring committed:
feat(labware-library): Support more pipettes in custom labware test protocol (#5084), and use opentrons_96_tiprack_20ul instead of 10ul

sfoster1 committed:
fix(app): disable module commands when protocol paused (#5209)

The module command cards displayed in the protocol run page have
historically allowed users to send module commands (like setting
temperature) when the protocol was active but paused. That used to work
essentially as a side effect: modules didn't respect pausing, so you
could tell them to do things while the protocol was paused.

Now that modules respect pausing, sending module commands when the
protocol is paused results in the modules not doing anything.

The science and UX intent of controlling modules from the protocol run
page is to implement pre-protocol preparation (like preheat) or
post-protocol behavior (like keeping some incubation going), not to
control the modules when paused - and in fact, when more advanced module
control behaviors land in protocol designer or in the python api,
controlling modules when the protocol is paused could break protocol
execution.

This PR changes the module command cards to only work when the protocol
is not paused or running.

When the module cards are disabled, the action button is now displayed but inaccessible, with an explanatory tooltip on hover.

Co-authored-by: Mike Cousins <mike@opentrons.com>

Commit:
refactor(robot-server): add /motors endpoints and tests (#5242)

* add /motors endpoints and tests

* create motor models from opentrons.hardware_control.types.Axis

* fix transform_validation_error_to_json_api_errors. error['loc'] may contain non strings.

closes #5192

shlokamin committed:
inject command maps into v4 executor

shlokamin committed:
reset mocks after each test run

shlokamin committed:
move JsonCommand enum into constants

shlokamin committed:
change error message

shlokamin commented on api/tests/opentrons/test_execute.py:
```diff
@@ -82,6 +82,8 @@ def emit_runlog(entry):
 
+# TODO IMMEDIATELY: test v4 fixture
+
 
```
looks like a test using a v4 fixture needs to be added

shlokamin commented on api/tests/opentrons/protocols/test_parse.py:
```diff
@@ -194,6 +196,10 @@ def test_validate_json(get_json_protocol_fixture, get_labware_fixture):
+    # TODO IMMEDIATELY: v4 fixture here
+    # v4 = get_json_protocol_fixture('4', 'someV4ProtocolTODO')
+    # assert validate_json(v4) == 4
+
```
same thing here, looks like a test using a v4 fixture needs to be added

IanLondon commented on api/tests/opentrons/protocols/test_parse.py:
```diff
@@ -194,6 +196,10 @@ def test_validate_json(get_json_protocol_fixture, get_labware_fixture):
     v3 = get_json_protocol_fixture('3', 'testAllAtomicSingleV3')
     assert validate_json(v3) == 3
 
+    # TODO IMMEDIATELY: v4 fixture here
```
we should add a v4 fixture and make sure it executes without crashing. Ideally it would do a gcode snapshot or something like that to make sure it's translated into PAPI calls exactly as we expect, but for now an integration smoke test would be nice.

IanLondon commented on api/tests/opentrons/test_execute.py:
```diff
@@ -82,6 +82,8 @@ def emit_runlog(entry):
         'Dropping tip into A1 of Trash on 12'
     ]
 
+# TODO IMMEDIATELY: test v4 fixture
```
same comment here as I made before

IanLondon reviewed (approved):
looks good, have comments about additional tests using the v4 fixture. Approved pending QA on robot :smile:

sfoster1 commented on api/src/opentrons/protocol_api/execute_v3.py:
```diff
@@ -77,8 +77,7 @@ def _get_location_with_offset(loaded_labware: Dict[str, labware.Labware],
 
-def _delay(
-        context, protocol_data, instruments, loaded_labware, params) -> None:
+def _delay(context, params) -> None:
```
type annotations for arguments?

shlokamin replied:
Yeah so this branch was actually based off of Ian's v3 executor refactor branch. I completely forgot that there were changes made to the v3 executor at all. I was trying to avoid touching the v3 executor for that reason, but since this PR already does, may as well address these changes too. Is that okay @IanLondon ?

sfoster1 commented on api/src/opentrons/protocol_api/execute_v3.py:
```diff
@@ -146,12 +139,11 @@ def _touch_tip(
     pipette = instruments[pipette_id]
     slot = params['slot']
-    if slot not in [str(s+1) for s in range(12)]:
+    if slot not in [str(s + 1) for s in range(12)]:
```
would probably rather this check on a deck definition, or the keys in the associated protocol context's `deck` object

sfoster1 commented on api/src/opentrons/protocol_api/execute_v3.py:
```diff
@@ -168,27 +160,46 @@ def _move_to_slot(
         minimum_z_height=params.get('minimumZHeight'))
 
 
+dispatcher_map: Dict[Any, Any] = {
```
this could probably use the nice command enums as keys, though from talking previously it might want to stay `Any` for values since all the functions are different

sfoster1 commented on api/src/opentrons/protocol_api/execute_v3.py:
```diff
@@ -168,27 +160,46 @@ def _move_to_slot(
-        "moveToSlot": _move_to_slot
-    }
+
+    pipette_command_list = [
```
This could involve enums a bit more

sfoster1 commented on api/src/opentrons/protocol_api/module_geometry.py:
```diff
@@ -515,6 +515,15 @@ def load_module(
 def resolve_module_model(module_name: str) -> ModuleModel:
     """ Turn any of the supported load names into module model names """
+
+    model_map: Mapping[str, ModuleModel] = {
```
[nitpitck] i think this should be a computed constant in this file, that could be something like
```
MODELS_BY_NAME = { e.value: e for e in itertools.chain(list(MagneticModuleModel), list(TemperatureModuleModel), list(ThermocyclerModuleModel)}
```

shlokamin replied:
I agree. I just tried this, and unfortunately it seems like mypy isn't smart enough to realize that `MODELS_BY_NAME ` is of type `Mapping[str, ModuleModel]`, but rather thinks it's of type `Mapping[str, _ProvideLookup]`.

Not sure how to get around this so I'll leave it for now. If you have any ideas lmk tho

sfoster1 reviewed (changes requested):
Couple little change requests but the structure looks good

shlokamin committed:
add type annotations to protocol context

shlokamin committed:
add type annotations for params

shlokamin committed:
add type annotations for instruments

shlokamin committed:
add type annotations for loaded labware

shlokamin committed:
add type annotations for set flow rate

shlokamin committed:
use JSON command enum in v3 executor

shlokamin committed:
use protocol context deck dict instead of static generated list

IanLondon committed:
fix(protocol-designer): fix bug with selecting magnet > disengage step (#5257)

Closes #5247

sfoster1 committed:
refactor(api): fullstep during plunger unstick (#5245)

It turns out that bumping the current isn't enough to unstick pipettes
that have been on the shelf for a really long time. We also need to
configure the smoothie to drive the pipette axes at full steps.

To do this, we introduce some new gcodes and wrap motion commands in
them.

This also restricts the move split / unsticks to pipette axes, since
only the pipette axes have configurable step ratios.

Also, bump split current to 1.75A

Closes #5168

sfoster1 committed:
refactor(api): increase gen2 multi flow rate (#5266)

These bump the flow rates from 5mm/s plunger speed to 10 mm/s plunger
speed per hw suggestion.

Closes #5264

IanLondon committed:
feat(protocol-designer): clear out engageHeight (#5254)

Clear out engageHeight from MAGNET steps when magnetic module model is
edited.

Closes #5225

IanLondon committed:
docs(shared-data): update v4 protocol schema - engageHeight description (#5259)

Closes #5227

IanLondon committed:
feat(shared-data): correct existing labware defs engageHeight (#5261)

Closes #5226

shlokamin committed:
use enum values as keys

Commit:
refactor(robot-server, api): robot server port /robot endpoints fast api (#5252)

* robot endpoint progress: home, light, and position implemented with unittests.

* add test for mount with z too low to aiohttp

* move api implemented and unit tested. moved camera into its own router.

* lint issues fixed

* move ThreadAsyncLock into hardware_control package.

* rename to threaded_async_lock and at it to hardware_control export

* use ThreadedAsyncLock in home and move handlers.

* fix lint errors

* add model validator tests. more clean up.

closes #5191

sfoster1 committed:
fix(api): make sure we move plunger to bottom after home (#5253)

This commit make a new function _do_plunger_home which homes a plunger
on a mount, if it exists, and then moves that plunger to its configured
bottom. This is now used both by API.home_plunger (which was where this
behavior was before) but also by a call to API.home() that involves
pipettes, ensuring that plungers are always moved to the bottom after
they are homed.

Closes #5164

sfoster1 committed:
fix(api): return tips to tipracks from the same height as drop tip (#5187)

Avoids an issue where the ejector would contact the tip rack body.

Closes #5186

IanLondon committed:
feat(protocol-designer): use different ranges for gen1 vs gen2 engageHeight (#5262)

Closes #5230

* introduce `.meta` to `HydratedFormData`

shlokamin committed:
compute model name constant

shlokamin committed:
update load module test to include all module names and aliases

shlokamin committed:
update labware test

shlokamin committed:
Revert "compute model name constant"

This reverts commit 9cb494079505e20751fe1ffe6cfeaabdfc780184.

shlokamin committed:
add json parse test for v4 fixture

shlokamin committed:
add json execute smoke test for v4 fixture

shlokamin committed:
fix lowercase name in test

shlokamin committed:
Merge branch 'api_add-v4-json-executor' into edge

shlokamin committed:
add require version for start set temperature command

shlokamin committed:
add start set temp and await temp to run log

sfoster1 reviewed (approved):
LGTM once qa approves. In the future we probably shouldn't merge edge back into longstanding branches though, it really makes a mess of the diff.

shlokamin commented:
Yeah that's my bad. I saw there was a conflict in the versioning notes so I thought I would merge and fix, but it really just made the tree unintelligible

shlokamin committed:
remove await temperature pause from the run log

shlokamin committed:
add await temp command to run log

shlokamin committed:
remove unnecessary msg argument
